apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs

```diff
@@ -0,0 +1,135 @@
+                            }
+                            catch (Exception e)
+                            {
+                                // If the commit actually aborted, calling Rollback will fail.
+                                // We don't want that or any other rollback exception to propogate,
+                                // it will not trigger the retry
+                                _connection.Logger.Warn("A rollback attempt failed on RetriableTransaction.RunAsync.CommitAttempt", e);
+                            }
+
+                            // Throw, the retry helper will know when to retry.
+                            throw;
+                        }
+                        finally
+                        {
+                            if (transaction != null)
+                            {
+                                // Let's make sure that the associated session is not released to the pool
+                                // because we'll be attempting to get a fresh transaction for this same session first.
+                                // If that doesn't work, then the session will be released to the pool and a new
+                                // session acquisition will be attempted.
+                                transaction.DisposeBehavior = DisposeBehavior.Detach;
+                                transaction.Dispose();
+                            }
+                        }
+                    }, "RetriableTransaction.RunAsync.CommitAttempt", _connection.Logger).ConfigureAwait(false);
+            }
+        }
+
+        internal async Task<TResult> ExecuteWithRetryAsync<TResult>(Func<Task<TResult>> fn, CancellationToken cancellationToken)
+        {
+            DateTime? overallDeadline = _options.CalculateDeadline(_clock);
+            TimeSpan retryDelay = _options.InitialDelay;
+            while (true)
+            {
+                try
+                {
+                    var result = await fn().ConfigureAwait(false);
+                    return result;
+                }
+                catch (SpannerException e) when (e.IsRetryable && !e.SessionExpired)
+                {
+                    // If there's a recommended retry delay specified on the exception
+                    // we should respect it.
+                    if (e.RecommendedRetryDelay.HasValue)
+                    {
+                        retryDelay = e.RecommendedRetryDelay.Value;
+                    }
+
+                    TimeSpan actualDelay = _options.Jitter(retryDelay);
+                    DateTime expectedRetryTime = _clock.GetCurrentDateTimeUtc() + actualDelay;
+                    if (expectedRetryTime > overallDeadline)
+                    {
+                        throw;
+                    }
+                    await _scheduler.Delay(actualDelay, cancellationToken).ConfigureAwait(false);
+                    retryDelay = _options.NextDelay(retryDelay);
```

Comment:
If we took a recommended delay, do we definitely want to double *that* rather than "the delay we'd have been using otherwise"?

Reply:
Apparently, having a retry backoff configuration is just for being on the very safe side, because on aborted the server will always recommend a delay.
I'm fine either way, but although I couldn't swear on it right now, I think I did this because it's what Java was doing.